core/commands/store/contracts/contracts.go

```diff
@@ -174,47 +189,126 @@ This command get contracts list based on role from the local node data store.`,
-			{
-				ContractId:              "737b6d38-5af1-4023-b219-196aadf4d3f0",
-				HostId:                  "16Uiu2HAmPTPCDrinViMyEzRGGcEJpc2VUH9bc46dU7sP2TzsbNnc",
-				RenterId:                "16Uiu2HAmCknnNaWa44X4kLRCq33B3zvBevRTLdo27qMhsszCwqdF",
-				Status:                  guardpb.Contract_DRAFT,
-				StartTime:               time.Now(),
-				EndTime:                 time.Now(),
-				NextEscrowTime:          time.Now(),
-				CompensationPaid:        0,
-				CompensationOutstanding: 15000,
-				UnitPrice:               10,
-				ShardSize:               500,
-				ShardHash:               "QmUX3GkfVQ8ARa79VE5HC6dxA5AtQQGaUTg1nbaqcAaYmp",
-				FileHash:                "QmAA3GkfVQ8ARa79VE5HC6dxA5AtQQGaUTg1nbaqcAaYm1",
-			},
-			{
-				ContractId:              "869675ac-c966-4808-83d7-1901d0449fb6",
-				HostId:                  "16Uiu2HAmPTPCDrinViMyEzRGGcEJpc2VUH9bc46dU7sP2TzsbNnc",
-				RenterId:                "16Uiu2HAmR6h5aamvwYDKYdp2Z3imCfHLRJnjB7VAYeab23AaZxSY",
-				Status:                  guardpb.Contract_CLOSED,
-				StartTime:               time.Now(),
-				EndTime:                 time.Now(),
-				NextEscrowTime:          time.Now(),
-				CompensationPaid:        100,
-				CompensationOutstanding: 300,
-				UnitPrice:               10,
-				ShardSize:               40,
-				ShardHash:               "QmTT3GkfVQ8ARa79VE5HC6dxA5AtQQGaUTg1nbaqcAaYm2",
-				FileHash:                "QmXx3GkfVQ8ARa79VE5HC6dxA5AtQQGaUTg1nbaqcAaYm3",
-			},
+		contracts, err := ListContracts(n.Repo.Datastore(), req.Arguments[0])
+		if err != nil {
+			return err
+		}
+		result := make([]*nodepb.Contracts_Contract, 0)
+		for _, c := range contracts {
+			b := false
+			for _, state := range states {
+				if state == c.Status {
+					continue
+				}
+			}
+			if !b {
+				continue
+			}
+			result = append(result, c)
 		}
-		return cmds.EmitOnce(res, &nodepb.Contracts{Contracts: data})
+		return cmds.EmitOnce(res, &nodepb.Contracts{Contracts: result})
 	},
 	Type: nodepb.Contracts{},
 }
+
+func getKey(role string) string {
+	var k string
+	if role == nodepb.ContractStat_HOST.String() {
+		k = hostContractsKey
+	} else if role == nodepb.ContractStat_RENTER.String() {
+		k = renterContractsKey
+	} else {
+		return "reserved"
+	}
+	return k
+}
+
+func Save(d datastore.Datastore, cs []*nodepb.Contracts_Contract, role string) error {
+	return ds.Save(d, getKey(role), &contractspb.Contracts{
+		Contracts: cs,
+	})
+}
+
+func ListContracts(d datastore.Datastore, role string) ([]*nodepb.Contracts_Contract, error) {
+	cs := &contractspb.Contracts{}
+	err := ds.Get(d, getKey(role), cs)
+	if err != nil && err != datastore.ErrNotFound {
+		return nil, err
+	}
+	return cs.Contracts, nil
+}
+
+func SyncContracts(ctx context.Context, n *core.IpfsNode, role string) error {
+	cfg, err := n.Repo.Config()
+	if err != nil {
+		return err
+	}
+	pk, err := n.Identity.ExtractPublicKey()
+	if err != nil {
+		return err
+	}
+	pkBytes, err := pk.Raw()
```

Comment:
Be careful, needs to use `RawFull`. Maybe put part of this into go-btfs-common ledger.go so you don't need to remember RawFull.

Reply:
nice catch, thanks.